IdentityCore/src/cache/metadata/accountMetadata/MSIDAccountMetadata.m

```diff
@@ -0,0 +1,216 @@
+    
+    self = [super init];
+    if (self)
+    {
+        _homeAccountId = homeAccountId;
+        _clientId = clientId;
+        _internalMap = [NSMutableDictionary new];
+        _signInState = MSIDAccountMetadataStateSignedIn;
+    }
+    return self;
+}
+
+#pragma mark - URL caching
+- (BOOL)setCachedURL:(NSURL *)cachedURL
+       forRequestURL:(NSURL *)requestURL
+       instanceAware:(BOOL)instanceAware
+               error:(NSError **)error
+{
+    _signInState = MSIDAccountMetadataStateSignedIn;
+    
+    if ([NSString msidIsStringNilOrBlank:cachedURL.absoluteString]
+        || [NSString msidIsStringNilOrBlank:requestURL.absoluteString])
+    {
+        if (error) *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInvalidInternalParameter, @"Either a target or request URL produces a nil string", nil, nil, nil, nil, nil, YES);
+        
+        return NO;
+    }
+    
+    NSString *urlMapKey = [self URLMapKey:instanceAware];
+    NSMutableDictionary *urlMap = _internalMap[urlMapKey];
+    if (!urlMap)
+    {
+        urlMap = [NSMutableDictionary new];
+        _internalMap[urlMapKey] = urlMap;
+    }
+    
+    urlMap[requestURL.absoluteString] = cachedURL.absoluteString;
+    return YES;
+}
+
+- (NSURL *)cachedURL:(NSURL *)requestURL instanceAware:(BOOL)instanceAware
+{
+    if (self.signInState != MSIDAccountMetadataStateSignedOut)
+    {
+        NSString *urlMapKey = [self URLMapKey:instanceAware];
+        NSDictionary *urlMap = _internalMap[urlMapKey];
+        
+        return [NSURL URLWithString:urlMap[requestURL.absoluteString]];
+    }
+    
+    return nil;
+}
+
+#pragma - Mark Signed out
+- (void)updateSignInState:(MSIDAccountMetadataState)state
+{
+    _signInState = state;
+    if (state == MSIDAccountMetadataStateSignedOut)
+    {
+        _internalMap = [NSMutableDictionary new];
+    }
+    
+}
+
+- (instancetype)initWithJSONDictionary:(NSDictionary *)json
+                                 error:(__unused NSError * __autoreleasing *)error
+{
+    if (!(self = [super init]))
+    {
+        return nil;
+    }
+    
+    if (!json)
+    {
+        MSID_LOG_WITH_CTX(MSIDLogLevelWarning,nil, @"Tried to decode an authority map item from nil json");
+        return nil;
+    }
+    
+    self->_clientId = [json msidStringObjectForKey:MSID_CLIENT_ID_CACHE_KEY];
+    self->_homeAccountId = [json msidStringObjectForKey:MSID_HOME_ACCOUNT_ID_CACHE_KEY];
+    self->_internalMap = [[json msidObjectForKey:MSID_AUTHORITY_MAP_CACHE_KEY ofClass:NSDictionary.class] mutableDeepCopy];
+    self->_signInState = [[json msidStringObjectForKey:MSID_SIGN_IN_STATE_CACHE_KEY] intValue];
+    
+    return self;
+}
+
+- (NSDictionary *)jsonDictionary
+{
+    NSMutableDictionary *dictionary = [NSMutableDictionary dictionary];
+    
+    dictionary[MSID_CLIENT_ID_CACHE_KEY] = self.clientId;
+    dictionary[MSID_HOME_ACCOUNT_ID_CACHE_KEY] = self.homeAccountId;
+    dictionary[MSID_AUTHORITY_MAP_CACHE_KEY] = _internalMap;
```

Comment:
nit: should this be a property for consistency?

Reply:
changed :)